[Navigation] Implement userInitiated in NavigateEvent

rwlbuis opened the pull request:
#### 28733f4c1575c6350918cc83e645f6a5f8e79c0e
<pre>
[Navigation] Implement userInitiated in NavigateEvent
<a href="https://bugs.webkit.org/show_bug.cgi?id=272146">https://bugs.webkit.org/show_bug.cgi?id=272146</a>

Reviewed by Alex Christensen.

Implement userInitiated in NavigateEvent.

Also correct isSameDocument determination when sending the push/replace/reload navigate event on fragment navigation since it is hardcoded to true:
<a href="https://html.spec.whatwg.org/multipage/browsing-the-web.html#scroll-to-fragid">https://html.spec.whatwg.org/multipage/browsing-the-web.html#scroll-to-fragid</a>:fire-a-push/replace/reload-navigate-event

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-anchor-userInitiated-expected.txt:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::innerDispatchNavigateEvent):

Canonical link: <a href="https://commits.webkit.org/277208@main">https://commits.webkit.org/277208@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c96338aa949233738870009971aa6e2722edcb5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46990 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26157 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49623 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49672 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43038 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49297 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31288 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23624 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38269 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47571 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23596 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40477 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19577 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20961 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41624 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5036 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/43370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42017 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51546 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/22010 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18368 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45562 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23292 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44551 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10378 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24067 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/23003 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rreno commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -424,7 +424,7 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
         info,
         canIntercept,
-        false, // FIXME: userInitiated
+        userInitiated,
```
Instead of passing this down as a parameter can we just query the document here in this functiion?

rwlbuis replied:
That is a good suggestion,I did not consider that. Will upload the new patch tomorrow, thanks!

rwlbuis replied:
Done, but also corrected the used API, i.e. hasRecentUserInteractionForNavigationFromJS was the wrong one.

TingPing commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -424,7 +425,7 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
         info,
         canIntercept,
-        false, // FIXME: userInitiated
+        UserGestureIndicator::processingUserGesture(),
```
Do click actions count as user gestures? Like on `a` elements?

rwlbuis replied:
Yes AFAIK (unlike a.click() which creates a simulated mouse event).

rreno replied:
Don't we need to pass the Document pointer in here? Otherwise we will never get a `true` result I think.

rreno replied:
Oh whoops that's not right. We can not pass in the document and still get true.

rwlbuis replied:
In fact it looks like passing the default document as null can/will return true due to the last statement in the function. But it does not seem the right thing indeed, I will try passing the document and check that the test still passes.

rwlbuis replied:
@rreno we came to the same conclusion, but it still looks like passing the document is the way to go.

rwlbuis committed:
[Navigation] Implement userInitiated in NavigateEvent
https://bugs.webkit.org/show_bug.cgi?id=272146

Reviewed by Alex Christensen.

Implement userInitiated in NavigateEvent.

Also correct isSameDocument determination when sending the push/replace/reload navigate event on fragment navigation since it is hardcoded to true:
https://html.spec.whatwg.org/multipage/browsing-the-web.html#scroll-to-fragid:fire-a-push/replace/reload-navigate-event

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-anchor-userInitiated-expected.txt:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::innerDispatchNavigateEvent):

Canonical link: https://commits.webkit.org/277208@main
